src/lstm_layer.cpp

```diff
@@ -1515,3 +1515,505 @@ void LSTM::preinit_layer() {
+            lstm_update_prev_hidden_states_worker(
+                this->lstm_states.mu_h_prior, this->lstm_states.var_h_prior,
+                input_delta_states.delta_mu, input_delta_states.delta_var, 0,
+                this->lstm_states.num_states, this->lstm_states.mu_h_prior,
+                this->lstm_states.var_h_prior);
+            lstm_update_prev_cell_states_worker(
+                this->lstm_states.mu_c_prior, this->lstm_states.var_c_prior,
+                this->lstm_states.jcb_ca, this->lstm_states.mu_o_ga,
+                input_delta_states.delta_mu, input_delta_states.delta_var, 0,
+                this->lstm_states.num_states, this->lstm_states.mu_c_prior,
+                this->lstm_states.var_c_prior);
+        }
+
+        // save posteriors for smoothing
+        this->lstm_states.mu_h_posts.push_back(this->lstm_states.mu_h_prior);
+        this->lstm_states.var_h_posts.push_back(this->lstm_states.var_h_prior);
+        this->lstm_states.mu_c_posts.push_back(this->lstm_states.mu_c_prior);
+        this->lstm_states.var_c_posts.push_back(this->lstm_states.var_c_prior);
+    }
+}
+
+void SLSTM::smoother(BaseTempStates &temp_states)
+/*
+ */
+{
+    int num_timestep = lstm_states.cov_cc.size();
+    int num_hidden_states = lstm_states.cov_cc[0].size();
+
+    lstm_states.mu_c_smooths.resize(num_timestep);
+    lstm_states.var_c_smooths.resize(num_timestep);
+    lstm_states.mu_h_smooths.resize(num_timestep);
+    lstm_states.var_h_smooths.resize(num_timestep);
+
+    for (auto &val : lstm_states.mu_c_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.var_c_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.mu_h_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.var_h_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+
+    // Initialize the last time step for smoothing
+    lstm_states.mu_c_smooths.back() = lstm_states.mu_c_posts.back();
+    lstm_states.var_c_smooths.back() = lstm_states.var_c_posts.back();
+    lstm_states.mu_h_smooths.back() = lstm_states.mu_h_posts.back();
+    lstm_states.var_h_smooths.back() = lstm_states.var_h_posts.back();
+
+    lstm_smoother_cell_states(num_timestep, lstm_states.cov_cc,
+                              lstm_states.mu_c_priors, lstm_states.var_c_priors,
+                              lstm_states.mu_c_posts, lstm_states.var_c_posts,
+                              lstm_states.mu_c_smooths,
+                              lstm_states.var_c_smooths);
+
+    lstm_smoother_hidden_states(
+        num_timestep, lstm_states.cov_hc, lstm_states.mu_c_priors,
+        lstm_states.var_c_priors, lstm_states.mu_h_posts,
+        lstm_states.var_h_posts, lstm_states.mu_c_smooths,
+        lstm_states.var_c_smooths, lstm_states.mu_h_smooths,
+        lstm_states.var_h_smooths);
+
+    // transfer h and c to the first time step of the next epoch
```

Comment:
Remove the dead code as much as possible. If not, put a TODO: in case where you want to go back to it later

Reply:
yes. I will